### ISSUE TITLE
Is Authentication possible just using the JSON data from the begin response from the FIDO server?

### ISSUE THREAD

**Issue by malateshk** [Yubico/libfido2#68]
Hi,

Assertion needs cred-id and for verification requires pubkey information from the developer manual page link :
https://developers.yubico.com/libfido2/Manuals/fido2-assert.html
{{{
$ echo assertion challenge | openssl sha256 -binary | base64 > assert_param
$ echo relying party >> assert_param
$ head -1 cred >> assert_param
$ tail -n +2 cred > pubkey
$ fido2-assert -G -i assert_param /dev/hidraw5 | fido2-assert -V pubkey es256
}}}

Is there a way to achieve authentication just using the response data from the demo server( from the authentication begin request-response) as we get only "public-key".

For example by using the below JSON response from the server: 

{"success":true,"request":{"requestId":"4qhlmPMGxi4R6bBABBchyHaeXXXXXXXXXXXXXX","publicKeyCredentialRequestOptions":{"challenge":"EXI31nxwsmN2FuHTXXXXXXXXXXXXXXX","rpId":"localhost","allowCredentials":[{"type":"public-key","id":"j4J3QfhoFus6BWQXXXXXXXXXXXXXXXXXXXXX"}],"userVerification":"preferred","extensions":{"appid":"https://localhost:8443"}},"username":"Demo"},"actions":{"finish":"https://localhost:8443/webauthn/api/v1/authenticate/finish"}}

Basically I am looking for a registration and authentication on FIDO server by using libfido2(C-library).

**Comment by martelletto** [comment]
Hi,

On 05/11/2019 17:26, malateshk wrote:
> $ fido2-assert -G -i assert_param /dev/hidraw5 | fido2-assert -V pubkey es256

This command does two things:

1) Talk to the authenticator to obtain an assertion;
2) Verifies the assertion using the specified public key.

Typically in a webauthn scenario, #2 happens on the RP, which has the
public key. For #1, all the attributes you need (challenge, RP id, and
credential id) are in the JSON blob below:

> {"success":true,"request":{"requestId":"4qhlmPMGxi4R6bBABBchyHaeXXXXXXXXXXXXXX","publicKeyCredentialRequestOptions":{"challenge":"EXI31nxwsmN2FuHTXXXXXXXXXXXXXXX","rpId":"localhost","allowCredentials":[{"type":"public-key","id":"j4J3QfhoFus6BWQXXXXXXXXXXXXXXXXXXXXX"}],"userVerification":"preferred","extensions":{"appid":"https://localhost:8443"}},"username":"Demo"},"actions":{"finish":"https://localhost:8443/webauthn/api/v1/authenticate/finish"}}

I hope to have answered your question.

-p.

**Comment by malateshk** [comment]
Hi Martelletto ,
Thank you for your response. 
Now the steps are clear to me. 

I used java demo server for registration and authentication via web browser and I got succeeded in doing so :

Now I tried the authentication from the libfido2 to the same Java demo server:
Here is the response I got from the demo server in begin step:

`Jason Response:{"success":true,"request":{"requestId":"8G8K3RDgrHBtYLyKRiWDYzsnFkMPioSfVIP84mqS0Ec","publicKeyCredentialRequestOptions":{"challenge":"JPDreQhxytR_AXs4Lbnc9HI7jq6BgE2P-8qtQK-hUZ0","rpId":"localhost","allowCredentials":[{"type":"public-key","id":"7TgXqmzpWyzyhZQNUOxnuFgPlr8_u-5k5El9JKemTXo7ieGTfnJ-8vsg21ROZEeGvYpQwu0LdsBU2rj2hayUPg"}],"userVerification":"preferred","extensions":{"appid":"https://localhost:8443"}},"username":"malatesh"},"actions":{"finish":"https://localhost:8443/webauthn/api/v1/authenticate/finish"}}`

- After the response, I attempted the get assertion as below:
1. echo "JPDreQhxytR_AXs4Lbnc9HI7jq6BgE2P-8qtQK-hUZ0" | openssl sha256 -binary | base64 > assert_param
2. echo "localhost" >> assert_param
3. echo "7TgXqmzpWyzyhZQNUOxnuFgPlr8_u-5k5El9JKemTXo7ieGTfnJ-8vsg21ROZEeGvYpQwu0LdsBU2rj2hayUPg" | base64 >> assert_param
4. fido2-assert -G -d -i assert_param /dev/hidraw3

**The 4th step supposed to get me the authdata, assertion signature and other details. But I get "fido2-assert: fido_dev_get_assert: FIDO_ERR_NO_CREDENTIALS".**

I would like to know, is there anything wrong in my approach to get the assertion details from the authenticator? I believe the required data should be available during earlier registration 
 through browser(make cred) .

Debug log of the get assertion is as below for reference:
{{{
**$ fido2-assert -G -d -i assert_param /dev/hidraw3**
client data hash:
  81 7f 28 f9 37 5e 07 b2 82 08 b9 47 5f f3 12 3d
  3f 18 b0 a6 df a2 15 f9 c5 cd 69 1d 86 ca bb e7
relying party id: localhost
credential id:
  37 54 67 58 71 6d 7a 70 57 79 7a 79 68 5a 51 4e
  55 4f 78 6e 75 46 67 50 6c 72 38 5f 75 2d 35 6b
  35 45 6c 39 4a 4b 65 6d 54 58 6f 37 69 65 47 54
  66 6e 4a 2d 38 76 73 67 32
tx: d=0x1496650, cmd=0x86, buf=0x1496650, count=8
  b0 ba 48 17 70 be 1d 19
rx: initiation frame at 0x7ffd8b1eefe0, len 64
  ff ff ff ff 86 00 11 b0 ba 48 17 70 be 1d 19 00
  96 00 14 02 05 01 02 05 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
tx: d=0x1496650, cmd=0x90, buf=0x1496b60, count=129
  02 a3 01 69 6c 6f 63 61 6c 68 6f 73 74 02 58 20
  81 7f 28 f9 37 5e 07 b2 82 08 b9 47 5f f3 12 3d
  3f 18 b0 a6 df a2 15 f9 c5 cd 69 1d 86 ca bb e7
  03 81 a2 62 69 64 58 39 37 54 67 58 71 6d 7a 70
  57 79 7a 79 68 5a 51 4e 55 4f 78 6e 75 46 67 50
  6c 72 38 5f 75 2d 35 6b 35 45 6c 39 4a 4b 65 6d
  54 58 6f 37 69 65 47 54 66 6e 4a 2d 38 76 73 67
  32 64 74 79 70 65 6a 70 75 62 6c 69 63 2d 6b 65
  79
rx: initiation frame at 0x7ffd8b1ee7d0, len 64
  00 96 00 14 90 00 01 2e 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
parse_cbor_reply: blob[0]=0x2e
fido_dev_get_assert_rx: adjust_assert_count
fido2-assert: fido_dev_get_assert: FIDO_ERR_NO_CREDENTIALS
}}}

**Comment by martelletto** [comment]
Hi,

> * After the response, I attempted the get assertion as below:
> 
> 1. echo "JPDreQhxytR_AXs4Lbnc9HI7jq6BgE2P-8qtQK-hUZ0" | openssl sha256 -binary | base64 > assert_param

The server is sending you a web-safe base64 encoded blob. You need to decode it, pipe it through OpenSSL to obtain the hash, and reencode it:

`$ echo "JPDreQhxytR_AXs4Lbnc9HI7jq6BgE2P-8qtQK-hUZ0=" | tr '_-' '/+' | openssl sha256 -binary | base64 > assert_param`

> 2. echo "localhost" >> assert_param

Please ensure this was the same RP id used for registration.

> 3. echo "7TgXqmzpWyzyhZQNUOxnuFgPlr8_u-5k5El9JKemTXo7ieGTfnJ-8vsg21ROZEeGvYpQwu0LdsBU2rj2hayUPg" | base64 >> assert_param

This is also a web-safe base64 encoded blob.

**Comment by malateshk** [comment]
Hi,

I tried the way as per your suggestion, but I am afraid still something wrong !!!!

Json Response from server
{"success":true,"request":{"requestId":"xBss0e1q4BiczQ8Nr4Th6A2OuCOCrle-1keXVPlaGus","publicKeyCredentialRequestOptions":{"challenge":"SlQ3EdHEecl6YV9QUoJlvK8rr_2a9YZMOksnw74eanI","rpId":"localhost","allowCredentials":[{"type":"public-key","id":"7TgXqmzpWyzyhZQNUOxnuFgPlr8_u-5k5El9JKemTXo7ieGTfnJ-8vsg21ROZEeGvYpQwu0LdsBU2rj2hayUPg"}],"userVerification":"preferred","extensions":{"appid":"https://localhost:8443"}},"username":"malatesh"},"actions":{"finish":"https://localhost:8443/webauthn/api/v1/authenticate/finish"}}

1.  **echo "SlQ3EdHEecl6YV9QUoJlvK8rr_2a9YZMOksnw74eanI" | tr '_-' '/+' | openssl sha256 -binary | base64 > assert_param**
2.  **echo "localhost" >> assert_param**   **=> The same rpId used for the registration**
3. **echo "7TgXqmzpWyzyhZQNUOxnuFgPlr8_u-5k5El9JKemTXo7ieGTfnJ-8vsg21ROZEeGvYpQwu0LdsBU2rj2hayUPg" | tr '_-' '/+' |  base64 >> assert_param**

**$ fido2-assert -G -d -i assert_param /dev/hidraw3**
client data hash:
  43 16 74 d7 cd f3 7b dc 5b cf 2b 7f af 86 c8 ce
  f4 73 03 94 4c 2e db 6b e1 9c 49 45 a8 3c c2 eb
relying party id: localhost
credential id:
  37 54 67 58 71 6d 7a 70 57 79 7a 79 68 5a 51 4e
  55 4f 78 6e 75 46 67 50 6c 72 38 2f 75 2b 35 6b
  35 45 6c 39 4a 4b 65 6d 54 58 6f 37 69 65 47 54
  66 6e 4a 2b 38 76 73 67 32
tx: d=0x10a2650, cmd=0x86, buf=0x10a2650, count=8
  9f aa 84 b5 21 88 a7 c0
rx: initiation frame at 0x7ffc2518d8a0, len 64
  ff ff ff ff 86 00 11 9f aa 84 b5 21 88 a7 c0 00
  96 00 24 02 05 01 02 05 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
tx: d=0x10a2650, cmd=0x90, buf=0x10a2b60, count=129
  02 a3 01 69 6c 6f 63 61 6c 68 6f 73 74 02 58 20
  43 16 74 d7 cd f3 7b dc 5b cf 2b 7f af 86 c8 ce
  f4 73 03 94 4c 2e db 6b e1 9c 49 45 a8 3c c2 eb
  03 81 a2 62 69 64 58 39 37 54 67 58 71 6d 7a 70
  57 79 7a 79 68 5a 51 4e 55 4f 78 6e 75 46 67 50
  6c 72 38 2f 75 2b 35 6b 35 45 6c 39 4a 4b 65 6d
  54 58 6f 37 69 65 47 54 66 6e 4a 2b 38 76 73 67
  32 64 74 79 70 65 6a 70 75 62 6c 69 63 2d 6b 65
  79
rx: initiation frame at 0x7ffc2518d090, len 64
  00 96 00 24 90 00 01 2e 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
parse_cbor_reply: blob[0]=0x2e
fido_dev_get_assert_rx: adjust_assert_count
fido2-assert: fido_dev_get_assert: FIDO_ERR_NO_CREDENTIALS

**A sample code to achieve the reg/authentication in real time client & demo server environment in example folder would be a great help, just like the java-webauthn-server demo.**

**Comment by martelletto** [comment]
Hi,

> 1. **echo "SlQ3EdHEecl6YV9QUoJlvK8rr_2a9YZMOksnw74eanI" | tr '_-' '/+' | openssl sha256 -binary | base64 > assert_param**

You need to base64-decode the blob before passing it to OpenSSL. Unfortunately, base64-decoding a web-safe blob on the command-line requires some work. You could do something like this:

```bash
#!/bin/bash -e

decode_websafe_b64() {
        local unescaped=$(echo "$1" | tr '_-' '/+')
        local n=$(echo -n "${unescaped}" | wc -c)
        local npadding=$((${n} % 4))
        if [[ ${npadding} -gt 2 ]]; then
                npadding=1 # ugh
        fi
        local padding=$(yes "=" | head -n${npadding} | tr -d '\n')
        echo "${unescaped}${padding}" | base64 -d
}

decode_websafe_b64 $1

```
Please don't rely on the code above. It is something I put together in 5 minutes to illustrate what needs to be done.

You could then do:

`./websafe_to_b64 <challenge> | openssl sha256 -binary | base64 > assert_param 
`

> 3. **echo "7TgXqmzpWyzyhZQNUOxnuFgPlr8_u-5k5El9JKemTXo7ieGTfnJ-8vsg21ROZEeGvYpQwu0LdsBU2rj2hayUPg" | tr '_-' '/+' |  base64 >> assert_param**

Same here.

> **A sample code to achieve the reg/authentication in real time client & demo server environment in example folder would be a great help, just like the java-webauthn-server demo.**

A robust example would require libcurl (or similar), a JSON library, and a proper web-safe base64 decoder. PRs welcome. :)

**Comment by martelletto** [comment]
Closing as the issue does not relate to a problem in the library. Feel free to contact me directly for support, if needed. Glad to help!
